File: drawing-java/lines.js
class TrendLine {
    constructor(startPoint, endPoint) {
        this.startPoint = startPoint;
        this.endPoint = endPoint;
    }

    adjustStartPoint(newStartPoint) {
        this.startPoint = newStartPoint;
    }

    adjustEndPoint(newEndPoint) {
        this.endPoint = newEndPoint;
    }

    draw(context) {
        context.beginPath();
        context.moveTo(this.startPoint.x, this.startPoint.y);
        context.lineTo(this.endPoint.x, this.endPoint.y);
        context.stroke();
    }

    getExtendedPoints() {
        return [this.startPoint, this.endPoint];
    }
}

class Ray extends TrendLine {
    getExtendedPoints() {
        const angle = Math.atan2(
            this.endPoint.y - this.startPoint.y,
            this.endPoint.x - this.startPoint.x
        );
        const length = 10000;
        
        const extendedEndPoint = {
            x: this.startPoint.x + Math.cos(angle) * length,
            y: this.startPoint.y + Math.sin(angle) * length
        };

        return [this.startPoint, extendedEndPoint];
    }

    draw(context) {
        const points = this.getExtendedPoints();
        context.beginPath();
        context.moveTo(points[0].x, points[0].y);
        context.lineTo(points[1].x, points[1].y);
        context.stroke();
    }
}

class ExtendedLine {
    constructor(startPoint, endPoint) {
        this.startPoint = startPoint;
        this.endPoint = endPoint;
        this.calculateAngle();
    }

    calculateAngle() {
        const dx = this.endPoint.x - this.startPoint.x;
        const dy = this.endPoint.y - this.startPoint.y;
        this.angleRad = Math.atan2(dy, dx);
    }

    adjustStartPoint(newStartPoint) {
        this.startPoint = newStartPoint;
        this.calculateAngle();
    }

    adjustEndPoint(newEndPoint) {
        this.endPoint = newEndPoint;
        this.calculateAngle();
    }

    getExtendedPoints(xRange = [-10000, 10000]) {
        const startX = xRange[0];
        const endX = xRange[1];
        const startY = this.startPoint.y + (startX - this.startPoint.x) * Math.tan(this.angleRad);
        const endY = this.startPoint.y + (endX - this.startPoint.x) * Math.tan(this.angleRad);
        return [
            { x: startX, y: startY },
            { x: endX, y: endY }
        ];
    }
}

class TrendAngle {
    constructor(startPoint, endPoint) {
        this.startPoint = startPoint;
        this.endPoint = endPoint;
    }

    calculateAngle() {
        const dx = this.endPoint.x - this.startPoint.x;
        const dy = this.endPoint.y - this.startPoint.y;
        const angleRad = Math.atan2(dy, dx);
        return angleRad * (180 / Math.PI);
    }

    getAngleText() {
        const angle = this.calculateAngle();
        return `${angle.toFixed(2)}°`;
    }
}

class HorizontalLine {
    constructor(yValue) {
        this.yValue = yValue;
    }

    adjustYValue(newYValue) {
        this.yValue = newYValue;
    }

    getExtendedPoints(xRange = [-10000, 10000]) {
        return [
            { x: xRange[0], y: this.yValue },
            { x: xRange[1], y: this.yValue }
        ];
    }
}

class VerticalLine {
    constructor(xValue) {
        this.xValue = xValue;
    }

    adjustXValue(newXValue) {
        this.xValue = newXValue;
    }

    getExtendedPoints(yRange = [-10000, 10000]) {
        return [
            { x: this.xValue, y: yRange[0] },
            { x: this.xValue, y: yRange[1] }
        ];
    }
}

class CrossLine {
    constructor(x, y) {
        this.x = x;
        this.y = y;
    }

    moveX(newX) {
        this.x = newX;
    }

    moveY(newY) {
        this.y = newY;
    }

    getPoints() {
        return { x: this.x, y: this.y };
    }
}

export {
    TrendLine,
    Ray,
    ExtendedLine,
    TrendAngle,
    HorizontalLine,
    VerticalLine,
    CrossLine
};

File: drawing-logic/lines.py
import math

class TrendLine:
    def __init__(self, start_point, end_point):
        self.start_point = start_point
        self.end_point = end_point

    def adjust_start_point(self, new_start_point):
        self.start_point = new_start_point

    def adjust_end_point(self, new_end_point):
        self.end_point = new_end_point

class Ray:
    def __init__(self, start_point, end_point):
        self.start_point = start_point
        self.end_point = end_point

    def adjust_start_point(self, new_start_point):
        self.start_point = new_start_point

    def adjust_end_point(self, new_end_point):
        self.end_point = new_end_point

class InfoRay:
    def __init__(self, start_point, end_point):
        self.start_point = start_point
        self.end_point = end_point

    def adjust_start_point(self, new_start_point):
        self.start_point = new_start_point

    def adjust_end_point(self, new_end_point):
        self.end_point = new_end_point

class ExtendedLine:
    def __init__(self, start_point, end_point):
        self.start_point = start_point
        self.end_point = end_point
        self.calculate_angle()

    def calculate_angle(self):
        x1, y1 = self.start_point
        x2, y2 = self.end_point
        self.angle_rad = math.atan2(y2 - y1, x2 - x1)

    def adjust_start_point(self, new_start_point):
        self.start_point = new_start_point
        self.calculate_angle()

    def adjust_end_point(self, new_end_point):
        self.end_point = new_end_point
        self.calculate_angle()

    def get_extended_points(self, x_range=(-10000, 10000)):
        x1, y1 = self.start_point
        start_x = x_range[0]
        end_x = x_range[1]
        start_y = y1 + (start_x - x1) * math.tan(self.angle_rad)
        end_y = y1 + (end_x - x1) * math.tan(self.angle_rad)
        return (start_x, start_y), (end_x, end_y)


class TrendAngle:
    def __init__(self, start_point, end_point):
        self.start_point = start_point
        self.end_point = end_point

    def adjust_start_point(self, new_start_point):
        self.start_point = new_start_point

    def adjust_end_point(self, new_end_point):
        self.end_point = new_end_point

    def calculate_angle(self):
        x1, y1 = self.start_point
        x2, y2 = self.end_point
        angle_rad = math.atan2(y2 - y1, x2 - x1)
        angle_deg = math.degrees(angle_rad)
        return angle_deg

    def get_angle_text(self):
        angle = self.calculate_angle()
        return f"{angle:.2f}°"

class HorizontalLine:
    def __init__(self, y_value):
        self.y_value = y_value

    def adjust_y_value(self, new_y_value):
        self.y_value = new_y_value

    def get_extended_points(self, x_range=(-10000, 10000)):
        start_x = x_range[0]
        end_x = x_range[1]
        return (start_x, self.y_value), (end_x, self.y_value)

class HorizontalRay:
    def __init__(self, start_point, end_point):
        self.start_point = start_point
        self.end_point = end_point

    def adjust_start_point(self, new_start_point):
        self.start_point = new_start_point

    def adjust_end_point(self, new_end_point):
        self.end_point = new_end_point

class VerticalLine:
    def __init__(self, x_value):
        self.x_value = x_value

    def adjust_x_value(self, new_x_value):
        self.x_value = new_x_value

    def get_extended_points(self, y_range=(-10000, 10000)):
        start_y = y_range[0]
        end_y = y_range[1]
        return (self.x_value, start_y), (self.x_value, end_y)
class VerticalLine:
    """Represents a vertical line that extends infinitely in both directions."""
    def __init__(self, x):
        self.x = x

    def draw(self):
        print(f"Drawing vertical line at x = {self.x}")

    def move_x(self, new_x):
        self.x = new_x

class CrossLine:
    """Represents a cross line that extends infinitely in all directions."""
    def __init__(self, x, y):
        self.x = x
        self.y = y

    def draw(self):
        print(f"Drawing cross line at ({self.x}, {self.y})")

    def move_x(self, new_x):
        self.x = new_x

    def move_y(self, new_y):
        self.y = new_y